portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,19 +14,97 @@
 import javax.servlet.http.HttpServletResponse;
+import java.util.List; 
+import java.util.ArrayList;
+import java.util.regex.Pattern;
+import java.util.regex.Matcher;
+import java.util.HashMap;
+import java.util.Map;
 
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
+  private HashMap<String, HashMap<String,Object>> information;  
+  private List<String> messages;
+
+  @Override
+  public void init(){
+      messages = new ArrayList<>();
+      information = new HashMap<String, HashMap<String,Object>>();
+  }
+
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello Thomas!</h1>");
+    
+    String json = convertToJsonUsingGson(information);
+    response.setContentType("application/json;");
+    response.getWriter().println(json);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException{
+      
+      information = checkValidity(request);
+
+      if (((Boolean) information.get("nameInfo").get("error") == true) || 
+          ((Boolean) information.get("messageInfo").get("error") == true)){
+          information.get("messageInfo").put("history", messages);
+          response.sendRedirect("/index.html#comment-page");
+      } else {
+          String message = (String) information.get("messageInfo").get("message");
+          messages.add(message);
+          information.get("messageInfo").put("history", messages);
+          response.sendRedirect("/index.html#comment-page");
+      }
+  }
+
+  private String convertToJsonUsingGson(HashMap<String, HashMap<String,Object>> messages) {
+    Gson gson = new Gson();
+    String json = gson.toJson(messages);
+    return json;
+  }
+  private HashMap<String, HashMap<String,Object>> checkValidity(HttpServletRequest request){
+
+    HashMap<String, HashMap<String,Object>> info = new HashMap<>();
+    HashMap<String, Object> message = new HashMap<>();
+    HashMap<String, Object> name = new HashMap<>();
+
+    String nameStr = request.getParameter("name");
+    String messageStr = request.getParameter("message-content");
+    
+    name.put("name", nameStr);
+    message.put("message", messageStr);
+    System.out.println(info);
+    if (!validateName(nameStr)){
+        name.put("error", true);
+    } else {
+        name.put("error", false);
+    }
+    if (validateMessage(messageStr)){
+        message.put("error", true);
+    } else {
+        message.put("error", false);
+    }
+
+    info.put("nameInfo", name);
+    info.put("messageInfo", message);
+
+    return info;
+  }
+  private boolean validateName(String text){
+      String regex = "^[a-zA-Z ]+$";
+      Pattern pattern =  Pattern.compile(regex, Pattern.CASE_INSENSITIVE);
```

Comment:
Generally you want to avoid compiling your regular expression on every function call for performance reasons (see [here](https://www.baeldung.com/java-regex-pre-compile), for example). 

Better to pre-compile and save the pattern object elsewhere, then use "reset" every time you use the pattern object again.

Reply:
Thanks! I didn't know about this :)

Reply:
Fixed in PR #9